design-documents/import/refactor-product-import.md

```diff
@@ -0,0 +1,380 @@
+    ```
+        /** @var \Magento\Framework\DataObject\IdentityService */
+        private $identityService;
+    ``` 
+* In the method `\Magento\CatalogImportExport\Model\Import\Product\Option::_importData` the lines
+    ```
+        $nextOptionId = $this->_resourceHelper->getNextAutoincrement($this->_tables['catalog_product_option']);
+        $nextValueId = $this->_resourceHelper->getNextAutoincrement(
+            $this->_tables['catalog_product_option_type_value']
+        );
+    ```
+    must be changed to
+    ```
+        $nextOptionId = $this->identityService->generateId();
+        $nextValueId = $this->identityService->generateId();
+    ```
+
+### Considerations
+
+Traditionally Magento uses `int(10)` to store serial primary keys, which evaluates to 4 bytes of data per row. 
+On MySQL / MariaDB UUID are stored as `char(36)` which evaluates to 36 bytes of data per row.
+This means that we'd add 32 bytes of storage per row if we switched to UUID.
+
+For the same reason, querying the database for an entity that uses UUID primary keys by its id could be a bit slower 
+as the index lookup may take a little longer (depending on the type of index that is being used).
+
+Since a process can create new entities without the need to talk to the database, the performance of the import will not
+degrade in any way, which is definitely worth the trade-off. 
+
+### Estimation
+
+As displayed above, refactoring the classes is rather easy as we simply have to replace fetching the `AUTO_INCREMENT`
+by generation of a UUID in the importers and give control over identity generation to the resource
+models. Altering the database and migrating from serial id to UUID without losing referential integrity on the other 
+hand will be a complex task that requires good planning and lots of testing.
+
+I'd estimate roughly 5 to 8 man-days for the job. 
+
+## Conclusion
+
+Although using UUIDs instead of serial ids would be a great option, migrating existing data could become troublesome and
+the overall performance of using a `char` field for primary key needs to be benchmarked before even attempting to use
+UUIDs. Additionally this is probably a BIC, thus it is not an option for a patch release.
+
+Switching to sequences on the other hand brings stability and thread-safety at minimal cost. The changes that need to be 
+done are straight forward and will definitely not raise a BIC.
+
+Considering all aspects, I'd suggest switching to generic sequences. In the long run I'd suggest to support sequences in
+the declarative schema and let the database abstraction layer decide how these are implemented for arbitrary RDBMS (e.g. 
+native sequences in Oracle, PostgreSQL and MSSQL, generic sequence table in MySQL / MariaDB etc.). The `identity` 
+attribute of columns in the declarative schema could then be used to indicate which column is fed by the sequence and 
```

Comment:
For distributed systems, it will not be possible to scale this kind of solution, where for generating ids we are relying on the database itself; we need to control at the program level how we want to generate the IDs. Twitter was using similar sequence for MySQL using AUTO_INCREMENT; but they realized it was not possible to scale this approach.

I would suggest looking into Twitter Snowflake 
https://github.com/twitter-archive/snowflake/tree/snowflake-2010

Another article around this subject is this
https://www.callicoder.com/distributed-unique-id-sequence-number-generator/

Although currently many of Magento customers running on single-node database, but in the cloud we would need to support large scale bulk operations in distributed style, in which case we would be able to allow multiple threads pushing the data; so an optimized service that can handle this scale would be better approach; which doesn't rely on hitting database for Ids.

Snowflake uses 8 byte unsigned integers, currently Magento using 4 byte unsigned integer, not-sure whether it will be BIC? but the idea is to centralize the ID generation and take it outside of database itself; the service should be similarly scalable as other services to support the load.